File: app/src/main/java/org/cerion/tasklist/dialogs/TaskListDialogFragment.java
package org.cerion.tasklist.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.widget.EditText;

import org.cerion.tasklist.data.AppDatabase;
import org.cerion.tasklist.data.TaskList;
import org.cerion.tasklist.data.TaskListDao;

import androidx.fragment.app.DialogFragment;

public class TaskListDialogFragment extends DialogFragment {

    private static final String TAG = TaskListDialogFragment.class.getSimpleName();
    private static final String TYPE = "type";
    private static final String LISTID = "id";
    private static final String LISTNAME = "name";

    public static final int TYPE_ADD = 0;
    public static final int TYPE_RENAME = 1;

    public static TaskListDialogFragment newInstance(int type, TaskList list) {
        TaskListDialogFragment frag = new TaskListDialogFragment();

        Bundle args = new Bundle();
        args.putInt(TYPE, type);
        if(list != null) {
            args.putString(LISTID,list.id);
            args.putString(LISTNAME,list.title);
        }
        frag.setArguments(args);

        return frag;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder alert = new AlertDialog.Builder(getActivity());
        final EditText edittext= new EditText(getActivity());
        alert.setTitle("Task List Name");
        alert.setView(edittext);

        Bundle bundle = getArguments();
        int type = bundle.getInt(TYPE);
        if(type == TYPE_ADD) {


            alert.setPositiveButton("Save", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int whichButton) {
                    String name = edittext.getText().toString();

                    TaskListDao db = AppDatabase.getInstance(getActivity()).taskListDao();
                    TaskList update = new TaskList(name);
                    db.add(update);

                    ((TaskListsChangedListener) getActivity()).onTaskListsChanged(update);
                }
            });

        }
        else if(type == TYPE_RENAME) {
            final String listId = bundle.getString(LISTID);
            final String listName = bundle.getString(LISTNAME);
            edittext.setText(listName);

            alert.setPositiveButton("Save", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int whichButton) {
                    String newName = edittext.getText().toString();
                    Log.d(TAG, "Rename " + listName + " to " + newName);

                    TaskList update = new TaskList(listId, newName, true);
                    TaskListDao db = AppDatabase.getInstance(getActivity()).taskListDao();
                    db.update(update);

                    ((TaskListsChangedListener) getActivity()).onTaskListsChanged(update);
                }
            });
        }

        alert.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                Log.d(TAG,"onCancel");
            }
        });

        return alert.create();
    }
}

File: app/src/main/java/org/cerion/tasklist/ui/dialogs/TaskListDialogFragment.kt
package org.cerion.tasklist.ui.dialogs

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.util.Log
import android.widget.EditText
import androidx.fragment.app.DialogFragment
import org.cerion.tasklist.database.AppDatabase
import org.cerion.tasklist.database.TaskList
import org.cerion.tasklist.database.getById
import java.util.*

class TaskListDialogFragment : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val alert = AlertDialog.Builder(activity)
        val edittext = EditText(activity)
        alert.setTitle("Task List Name")
        alert.setView(edittext)

        val bundle = arguments
        val type = bundle!!.getInt(TYPE)
        if (type == TYPE_ADD) {

            alert.setPositiveButton("Save") { _, _ ->
                val name = edittext.text.toString()

                val db = AppDatabase.getInstance(requireContext())!!.taskListDao()
                val update = TaskList(AppDatabase.generateTempId(), name)
                db.add(update)

                // TODO not sure if this method is prefered or onActivityResult is better
                if (targetFragment is TaskListsChangedListener)
                    (targetFragment as TaskListsChangedListener).onTaskListsChanged(update)
            }
        } else if (type == TYPE_RENAME) {
            val listId = bundle.getString(LISTID)
            val listName = bundle.getString(LISTNAME)
            edittext.setText(listName)

            alert.setPositiveButton("Save") { _, _ ->
                val newName = edittext.text.toString()
                Log.d(TAG, "Rename $listName to $newName")

                val db = AppDatabase.getInstance(requireActivity())!!.taskListDao()
                val list = db.getAll().getById(listId!!)!!
                list.title = newName
                list.updated = Date()
                db.update(list)

                val parent = targetFragment as TaskListsChangedListener
                parent.onTaskListsChanged(list)
            }
        }

        alert.setNegativeButton("Cancel") { _, _ -> Log.d(TAG, "onCancel") }

        return alert.create()
    }

    companion object {

        private val TAG = TaskListDialogFragment::class.java.simpleName
        private const val TYPE = "type"
        private const val LISTID = "id"
        private const val LISTNAME = "name"

        private const val TYPE_ADD = 0
        private const val TYPE_RENAME = 1

        fun getAddInstance(): TaskListDialogFragment {
            val frag = TaskListDialogFragment()

            val args = Bundle()
            args.putInt(TYPE, TYPE_ADD)
            frag.arguments = args

            return frag
        }

        fun getRenameInstance(list: TaskList): TaskListDialogFragment {
            val frag = TaskListDialogFragment()
            val args = Bundle()
            args.putInt(TYPE, TYPE_RENAME)
            args.putString(LISTID, list.id)
            args.putString(LISTNAME, list.title)
            frag.arguments = args

            return frag
        }
    }
}
